### PR TITLE
 Feature: input view handling.

### PR DESCRIPTION
**Pull request by 3lvis**

### CODE REVIEW

File: iOSDemo/EditViewController.swift [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+    lazy var keyboardAwareInputView: KeyboardAwareInputAccessoryView = {
+        let view = KeyboardAwareInputAccessoryView(withAutoLayout: true)
+        view.delegate = self
+
+        return view
+    }()
+
+    override var inputAccessoryView: UIView? {
+        return self.keyboardAwareInputView
```
Comment (elland) [inline iOSDemo/EditViewController.swift]: I want to remove this boilerplate here.

**Comment by elland** [comment]
Feedback? Ideas how to improve?

File: Sources/KeyboardAwareInputViewProtocol.swift [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    /// This method is called every time the keyboard moves and sends back the keyboard origin y coordinate point.
+    /// Use this to keep your view always above the keyboard.
+    ///
+    /// - Also:
+    ///   - When the keyboard is offscreen, the point returned is 0.
+    ///
+    /// - Parameters:
+    ///   - inputView: The KeyboardAwareInputAccessoryView instance. You don't usually need it.
+    ///   - keyboardOriginYDistance: The distance from the top of the keyboard to the bottom of the screen.
+    func inputView(_ inputView: KeyboardAwareInputAccessoryView, shouldUpdatePosition keyboardOriginYDistance: CGFloat)
```
Comment (elland) [inline Sources/KeyboardAwareInputViewProtocol.swift:30]: `keyboardOriginYDistance` is still a ridiculous name. 😭

File: iOSDemo/EditViewController.swift [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    }()
+
+    lazy var textInputView: UIView = {
+        let view = UIView(withAutoLayout: true)
+        view.backgroundColor = .magenta
+
+        return view
+    }()
+
+    // TODO: Find a way to make this work implicitly.
```
Comment (3lvis) [inline iOSDemo/EditViewController.swift]: This will trigger a warning in my app. Can you remove this comment?

**Comment by 3lvis** [comment]
@Elland If it's ready for reviewing drop the WIP from the title :D

**Comment by elland** [comment]
@3lvis ready for review and feedback not merging. I wanted to see if either you or @marijnschilling had any ideas how to improve the client side API.

**Comment by 3lvis** [comment]
I think we should improve the demo so it shows one of the scenarios where we would find this useful.

**Comment by elland** [comment]
Anything simple in mind?

**Comment by 3lvis** [comment]
Also `KeyboardAwareAccessoryViewDelegate` feels like a "responder" type of protocol, meaning a protocol that just informs me when something happens, meanwhile this protocol also requires me to have certain attributes. It's like a UIScrollViewDelegate would require you to have a `var isScrolling: Bool` variable in order to work. Maybe the name should sound more like a behavior that you're introducing, maybe just something like `KeyboardAware`.

**Comment by 3lvis** [comment]
@Elland Maybe a UITextField as accessory that would add into a tableView everytime you press enter?

**Comment by elland** [comment]
Although I agree with the sentiment, that was also the goal. To remove the mandatory property override and have it just work. Swift is not dynamic enough for that though. 

`KeyboardAware` on the other hand just doesn't tell me anything. 🤔

**Comment by 3lvis** [comment]
Did you end up doing this or should we close the PR?

**Comment by elland** [comment]
I think we can finish this. We're also using something similar for Token. 🤔

**Comment by elland** [comment]
@roberthein wanna take a look?

File: Sources/UIView+Sweet.swift [hunk 4]
```diff
@@ -25,10 +25,17 @@
         public func fillSuperview(with insets: UIEdgeInsets = UIEdgeInsets.zero) {
             guard let superview = self.superview else { return }
 
-            self.topAnchor.constraint(equalTo: superview.topAnchor, constant: insets.top).isActive = true
-            self.leftAnchor.constraint(equalTo: superview.leftAnchor, constant: insets.left).isActive = true
-            self.rightAnchor.constraint(equalTo: superview.rightAnchor, constant: -insets.right).isActive = true
-            self.bottomAnchor.constraint(equalTo: superview.bottomAnchor, constant: -insets.bottom).isActive = true
+            let constants = [
+                self.topAnchor.constraint(equalTo: superview.topAnchor, constant: insets.top),
```
Comment (roberthein) [inline Sources/UIView+Sweet.swift]: I think `constants` is a confusing name for an array of constraints.

File: Sources/KeyboardAwareInputViewProtocol.swift [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+             We then use it's negative value, as we're dealing with an inversed constraint. That's it.
+             */
+            guard let window = self.window else { return }
+            let constant = -(superview.frame.intersection(window.bounds).height)
+
+            self.delegate?.inputView(self, shouldUpdatePosition: constant)
+        } else {
+            super.observeValue(forKeyPath: keyPath, of: object, change: change, context: context)
+        }
+    }
```
Comment (roberthein) [inline Sources/KeyboardAwareInputViewProtocol.swift:120]: 💯

File: Sources/UIView+Sweet.swift [hunk 6]
```diff
@@ -25,10 +25,17 @@
...
+                self.leftAnchor.constraint(equalTo: superview.leftAnchor, constant: insets.left),
+                self.rightAnchor.constraint(equalTo: superview.rightAnchor, constant: -insets.right),
+                self.bottomAnchor.constraint(equalTo: superview.bottomAnchor, constant: -insets.bottom)
+            ]
+
+            constants.forEach { constant in
+                constant.priority = 999
+                constant.isActive = true
+            }
```
Comment (roberthein) [inline Sources/UIView+Sweet.swift]: Here you could get `constant.constant = 100` for example..

File: iOSDemo/EditViewController.swift [hunk 7]
```diff
@@ -0,0 +1,64 @@
+import UIKit
+import SweetUIKit
+
+class EditViewController: UIViewController, KeyboardAwareAccessoryViewDelegate {
+
```
Comment (roberthein) [inline iOSDemo/EditViewController.swift:5]: Good demo 👍

**Review by roberthein — APPROVED** [review]
🚀

**Comment by roberthein** [comment]
It is very similar for what we are doing in token, we should use this ;)

**Comment by 3lvis** [comment]
@Elland Might be good if you add a checklist of what's missing so anyone can help to wrap this up.

**Comment by elland** [comment]
@3lvis sounds good. I'll see if I can get around to doing this today or tomorrow. 👍

**Comment by elland** [comment]
@roberthein mind documenting usage in the Readme file?

**Comment by roberthein** [comment]
I added an entry in the README about the keyboard aware accessory view. Can we merge this now?

**Comment by elland** [comment]
I guess we can improve the Readme later, with some sample code, maybe even a gif. Solve conflicts and we merge?

**Review by roberthein — APPROVED** [review]
👍

**Pull request merged**